File: leetcode/1-100/29-divide-two-integers/bench/bit_shift.c
/*
 * LeetCode 29 — bit-shift long division, C mirror.
 * Algorithmic peer of bench/bit_shift.{kara,rs,py}. Same N, LCG inputs,
 * same truncating bit-shift divide, same sink formula.
 */
#include <stdio.h>
#include <stdint.h>

static int64_t divide(int64_t dividend, int64_t divisor) {
    const int64_t int_max = 2147483647;
    const int64_t int_min = -2147483648;
    if (dividend == int_min && divisor == -1) {
        return int_max;
    }
    int negative = (dividend < 0) != (divisor < 0);
    int64_t a = dividend < 0 ? -dividend : dividend;
    int64_t b = divisor < 0 ? -divisor : divisor;
    int64_t result = 0;
    while (a >= b) {
        int64_t temp = b;
        int64_t multiple = 1;
        while (a >= (temp << 1)) {
            temp <<= 1;
            multiple <<= 1;
        }
        a -= temp;
        result += multiple;
    }
    return negative ? -result : result;
}

int main(void) {
    const int64_t n = 5000000;
    int64_t state = 1;
    int64_t sum = 0;
    for (int64_t i = 0; i < n; i++) {
        state = (state * 1103515245 + 12345) % 2147483648;
        int64_t dividend = state - 1073741824;
        state = (state * 1103515245 + 12345) % 2147483648;
        int64_t divisor = (state % 2000) - 1000;
        if (divisor == 0) divisor = 1;
        sum += divide(dividend, divisor);
    }
    printf("%lld\n", (long long)sum);
    return 0;
}

File: leetcode/1-100/29-divide-two-integers/bit_shift.py
"""LeetCode #29: Divide Two Integers — bit-shift long division.

Algorithmic mirror of bit_shift.kara. Output format matches line-for-line so
the two can be diffed directly.

Note: this re-implements truncating division with shifts rather than using
Python's `//` (which floors toward negative infinity) — the whole point of the
kata is to divide without `*`, `/`, or `%`, and to truncate toward zero.
"""

from __future__ import annotations


def divide(dividend: int, divisor: int) -> int:
    int_max = 2147483647
    int_min = -2147483648

    # The sole result that overflows the signed 32-bit range — saturate.
    if dividend == int_min and divisor == -1:
        return int_max

    # Sign of the quotient is the XOR of the operand signs.
    negative = (dividend < 0) != (divisor < 0)

    a = abs(dividend)
    b = abs(divisor)

    result = 0
    while a >= b:
        # Largest shifted divisor that still fits under the remainder.
        temp = b
        multiple = 1
        while a >= (temp << 1):
            temp <<= 1
            multiple <<= 1
        a -= temp
        result += multiple

    return -result if negative else result


def main() -> None:
    print(divide(10, 3))            # 3
    print(divide(7, -3))            # -2
    print(divide(-2147483648, -1))  # 2147483647 (overflow → clamp)
    print(divide(-2147483648, 1))   # -2147483648
    print(divide(-2147483648, 2))   # -1073741824
    print(divide(2147483647, 1))    # 2147483647
    print(divide(0, 5))             # 0
    print(divide(1, 1))             # 1
    print(divide(-7, -3))           # 2
    print(divide(15, 4))            # 3
    print(divide(-15, 4))           # -3
    print(divide(1, -1))            # -1
    print(divide(-2147483648, -2))  # 1073741824


if __name__ == "__main__":
    main()
